q2cli/util.py

```diff
@@ -234,3 +275,119 @@ def get_plugin_manager():
             return pm
 
         return qiime2.sdk.PluginManager()
+
+
+def load_metadata(fp):
+    """ Turns a filepath into metadata if the path is either to metadata or an
+    artifact that can be viewed as metadata
+    """
+    import sys
+    import qiime2
+
+    try:
+        artifact = get_input(fp)
+    except ControlFlowException:
+        try:
+            return qiime2.Metadata.load(fp)
+        except Exception as e:
+            # Handle the case where we try to load a raw .qza as Metadata
+            # because we were missing the pugin(s) necessary to load it as an
+            # Artifact
+            if "'utf-8' codec can't decode byte" in str(e) and \
+                    qiime2.sdk.result.Result.peek(fp):
+                e = ValueError("It looks like you have an Artifact but are "
+                               "missing the plugin(s) necessary to load it.")
```

Comment:
This kind of circles around the original problem. Perhaps we refactor `get_input` to return more than just the result? And then call-sites can error as they need/want to?

Exceptions are technically something you can pass around as any other object.

Reply:
What do we gain from doing that though? Maybe I've just looked at this too much, so it makes too much sense to me now, but in get_input we raise a ControlFlowException if something we didn't anticipate happens, and we raise more informative errors if we do understand what happened. Why would we pass them around instead of just raising them and letting the call site handle them? What do we gain by returning them or something instead?